src/System.Net.Security/src/System/Net/_SecureChannel.cs

```diff
@@ -0,0 +1,1487 @@
+                {
+                    store = isMachineStore ? s_MyMachineCertStoreEx : s_MyCertStoreEx;
+                    if (store == null)
+                    {
+                        // NOTE: that if this call fails we won't keep track and the next time we enter we will try to open the store again.
+                        StoreLocation storeLocation = isMachineStore ? StoreLocation.LocalMachine : StoreLocation.CurrentUser;
+                        store = new X509Store(StoreName.My, storeLocation);
+                        try
+                        {
+                            // For app-compat We want to ensure the store is opened under the **process** account.
+                            try
+                            {
+                                WindowsIdentity.RunImpersonated(SafeAccessTokenHandle.InvalidHandle, () =>
+                                {
+                                    store.Open(OpenFlags.ReadOnly | OpenFlags.OpenExistingOnly);
+                                    GlobalLog.Print("SecureChannel::EnsureStoreOpened() storeLocation:" + storeLocation + " returned store:" + store.GetHashCode().ToString("x"));
+                                });
+                            }
+                            catch
+                            {
+                                throw;
+                            }
+
+                            if (isMachineStore)
+                                s_MyMachineCertStoreEx = store;
+                            else
+                                s_MyCertStoreEx = store;
+
+                            return store;
+                        }
+                        catch (Exception exception)
+                        {
+                            if (exception is CryptographicException || exception is SecurityException)
+                            {
+                                GlobalLog.Assert("SecureChannel::EnsureStoreOpened()", "Failed to open cert store, location:" + storeLocation + " exception:" + exception);
+                                return null;
+                            }
+                            if (Logging.On) Logging.PrintError(Logging.Web, SR.Format(SR.net_log_open_store_failed, storeLocation, exception));
+                            throw;
+                        }
+                    }
+                }
+            }
+            return store;
+        }
+
+        private static X509Certificate2 MakeEx(X509Certificate certificate)
+        {
+            Debug.Assert(certificate != null, "certificate != null");
+
+            if (certificate.GetType() == typeof(X509Certificate2))
+            {
+                return (X509Certificate2)certificate;
+            }
+
+            X509Certificate2 certificateEx = null;
+            try
+            {
+                if (certificate.Handle != IntPtr.Zero)
+                {
+                    certificateEx = new X509Certificate2(certificate.Handle);
```

Comment:
On Unix the association of the private key to the public certificate is done only in managed code, so the handle ctor will be guaranteed to lose that association.

But, on Unix you shouldn't ever encounter an object whose GetType() == typeof(X509Certificate); so maybe it's fine.